Mathlib/FieldTheory/JacobsonNoether.lean

```diff
@@ -0,0 +1,198 @@
+    use ha.choose
+    show a * ha.choose - ha.choose * a ≠ 0
+    simpa only [ne_eq, sub_eq_zero] using Ne.symm ha.choose_spec
+  -- We find a maximum natural number `n` such that `(a * x - x * a) ^ n b ≠ 0`.
+  obtain ⟨n, hn, hb⟩ : ∃ n, 0 < n ∧ (ad k D a)^[n] b ≠ 0 ∧ (ad k D a)^[n + 1] b = 0 := by
+    obtain ⟨m, -, hm2⟩ := exist_pow_eq_zero_of_le p ha insep
+    have h_exist : ∃ n, 0 < n ∧ (ad k D a)^[n + 1] b = 0 := ⟨p ^ m,
+      ⟨expChar_pow_pos D p m, by rw [hm2 (p ^ m + 1) (Nat.le_add_right _ _)]; rfl⟩⟩
+    classical
+    refine ⟨Nat.find h_exist, ⟨(Nat.find_spec h_exist).1, ?_, (Nat.find_spec h_exist).2⟩⟩
+    set t := (Nat.find h_exist - 1 : ℕ) with ht
+    by_cases h_pos : 0 < t
+    · convert (ne_eq _ _) ▸ not_and.mp (Nat.find_min h_exist (m := t) (by omega)) h_pos
+      omega
+    · suffices h_find: Nat.find h_exist = 1 by
+        rwa [h_find]
+      rw [not_lt, Nat.le_zero, ht, Nat.sub_eq_zero_iff_le] at h_pos
+      linarith [(Nat.find_spec h_exist).1]
+  -- We define `c` to be the value that we proved above to be non-zero.
+  set c := (ad k D a)^[n] b with hc_def
+  let _ : Invertible c := ⟨c⁻¹, inv_mul_cancel₀ hb.1, mul_inv_cancel₀ hb.1⟩
+  -- We prove that `c` commutes with `a`.
+  have hc : a * c = c * a := by
+    apply eq_of_sub_eq_zero
+    rw [← mulLeft_apply (R := k), ← mulRight_apply (R := k)]
+    suffices ad k D a c = 0 from by
+      rw [← this]; rfl
+    rw [ad_iterate_succ a b n, hb.2]
+  -- We now make some computation to obtain the final equation.
+  set d := c⁻¹ * a * (ad k D a)^[n - 1] b with hd_def
+  have hc': c⁻¹ * a = a * c⁻¹ := by
+    apply_fun (c⁻¹ * · * c⁻¹) at hc
+    rw [mul_assoc, mul_assoc, mul_inv_cancel₀ hb.1, mul_one, ← mul_assoc,
+      inv_mul_cancel₀ hb.1, one_mul] at hc
+    exact hc
+  have c_eq : a * (ad k D a)^[n - 1] b - (ad k D a)^[n - 1] b * a = c := by
+    rw [hc_def, ← Nat.sub_add_cancel hn, ← ad_iterate_succ]; rfl
+  have eq1 : c⁻¹ * a * (ad k D a)^[n - 1] b - c⁻¹ * (ad k D a)^[n - 1] b * a = 1 := by
+    simp_rw [mul_assoc, (mul_sub_left_distrib c⁻¹ _ _).symm, c_eq, inv_mul_cancel_of_invertible]
+  -- We show that `a` commutes with `d`.
+  have deq : a * d - d * a = a := by
+    nth_rw 3 [← mul_one a]
+    rw [hd_def, ← eq1, mul_sub, mul_assoc _ _ a, sub_right_inj, hc',
+      ← mul_assoc, ← mul_assoc, ← mul_assoc]
+  -- This then yields a contradiction.
+  apply_fun (a⁻¹ * · ) at deq
+  rw [mul_sub, ← mul_assoc, inv_mul_cancel₀ ha₀, one_mul, ← mul_assoc, sub_eq_iff_eq_add] at deq
+  obtain ⟨r, hr⟩ := exists_pow_mem_center_of_inseparable p d insep
+  apply_fun (· ^ (p ^ r)) at deq
+  rw [add_pow_expChar_pow_of_commute p r (Commute.one_left _) , one_pow,
+    conj_pow₀ ha₀, ← hr.comm, mul_assoc, inv_mul_cancel₀ ha₀, mul_one,
+    self_eq_add_left] at deq
+  exact one_ne_zero deq
+
+open Subring Algebra in
+/-- Jacobson-Noether theorem: For a non-commutative division algebra `D`
+  that is algebraic over a field `L`, if the center of
```

Comment:
Is this really saying that `L` is the center? It seems more general a priori to me.

Reply:
I think this is precisely [Algebra.IsCentral](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Central/Defs.html#Algebra.IsCentral): it says that the center of `D` coincides with the image of `L` in `D`.

Reply:
But now I am confused by the naming here. I would think `exists_separable_of_not_top` or `exists_separable_of_not_comm` We really need to train a model to generate declaration names.

Reply:
I see what you mean, but mathematically we would simply identify `L` with its image, so saying that an element is not in `top` really means that it is not central and I feel that `not_central` (or `not_comm`, see below) conveys more mathematical meening. Concerning `not_comm`, I have the slight feeling that this would refer to a couple of elements commuting with each other rather than an element commuting with everything, but I agree this is not very sound.

Reply:
It seems strange to me that we have `IsCentral` as a parameter and we use `not_central` in the name. Also, what is `mem` doing? I couldn't quite parse it. It would nice to have some AI trained to generate these names from the declaration itself.

Reply:
Either way 

bors d=faenuccio

Reply:
@Yu-Misaka OK, we're at the end. Following Matt's advice, I would like that your rename the two main theorems by
1. Removing "mem" (it applies to conclusion of the form "something belongs to...", which is not the case)
2. Changing `central` to `isCentral`.

Then I will merge it.

Reply:
Thank you so much for your work!

Reply:
Thank you for your help! It's been a pleasure working with you!

I've implemented the change;)

Reply:
Oh! I realized I was confused by the `of` since the **conclusion** is that you have a separable non central element.

Reply:
Right?

Reply:
Yep!

Reply:
Sure, no problem;-)